File: frontend/src/types/race.ts
export interface Horse {
  horse_id: string;
  name: string;
  sex?: string;
  birth_year?: number;
  father?: string;
  mother?: string;
  trainer?: string;
}

export interface Jockey {
  jockey_id: string;
  name: string;
  win_rate?: number;
  place_rate?: number;
  show_rate?: number;
}

export interface Entry {
  horse_number: number;
  frame_number?: number;
  horse_id: string;
  horse_name?: string;
  horse?: Horse;
  jockey_id?: string;
  jockey_name?: string;
  jockey?: Jockey;
  weight?: number;
  horse_weight?: number;
  weight_diff?: number;
  odds?: number;
  popularity?: number;
  result?: number;
  finish_time?: string;
  last_3f?: number;
}

export interface Race {
  race_id: string;
  date: string;
  course: string;
  race_number: number;
  race_name?: string;
  distance: number;
  track_type: string;
  weather?: string;
  condition?: string;
  grade?: string;
  entries?: Entry[];
}

export interface RaceListResponse {
  total: number;
  races: Race[];
}

export interface RaceDetailResponse extends Race {
  entries: Entry[];
}

File: backend/app/schemas/race.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field


class HorseBase(BaseModel):
    horse_id: str
    name: str
    sex: str
    birth_year: int


class HorseResponse(HorseBase):
    father: Optional[str] = None
    mother: Optional[str] = None
    mother_father: Optional[str] = None
    trainer: Optional[str] = None

    class Config:
        from_attributes = True


class JockeyBase(BaseModel):
    jockey_id: str
    name: str


class JockeyResponse(JockeyBase):
    win_rate: Optional[float] = None
    place_rate: Optional[float] = None
    show_rate: Optional[float] = None

    class Config:
        from_attributes = True


class EntryBase(BaseModel):
    horse_number: int
    frame_number: Optional[int] = None
    weight: Optional[float] = None
    horse_weight: Optional[int] = None
    weight_diff: Optional[int] = None
    odds: Optional[float] = None
    popularity: Optional[int] = None


class EntryResponse(EntryBase):
    id: int
    race_id: str
    horse_id: Optional[str] = None
    horse_name: Optional[str] = None
    jockey_id: Optional[str] = None
    jockey_name: Optional[str] = None
    result: Optional[int] = None
    finish_time: Optional[str] = None
    margin: Optional[str] = None
    corner_position: Optional[str] = None
    last_3f: Optional[float] = None

    class Config:
        from_attributes = True


class RaceBase(BaseModel):
    race_id: str
    date: date
    course: str
    race_number: int
    distance: int
    track_type: str


class RaceListItem(RaceBase):
    race_name: Optional[str] = None
    grade: Optional[str] = None

    class Config:
        from_attributes = True


class RaceDetail(RaceBase):
    race_name: Optional[str] = None
    weather: Optional[str] = None
    condition: Optional[str] = None
    grade: Optional[str] = None
    entries: list[EntryResponse] = []

    class Config:
        from_attributes = True


class RaceListResponse(BaseModel):
    total: int
    races: list[RaceListItem]


class ScrapeRaceRequest(BaseModel):
    date: str = Field(..., pattern=r"^\d{4}-\d{2}-\d{2}$", description="Date in YYYY-MM-DD format")
    course: Optional[str] = Field(None, description="Course name filter")


class ScrapeRaceResponse(BaseModel):
    status: str
    message: str
    races_count: int
    saved_count: int = 0
    races: list[dict] = []


class ScrapeDetailResponse(BaseModel):
    status: str
    saved: bool
    race: dict
